### PR TITLE
[SDK-2458] Implement Consumer Protection Preferences

### PR DESCRIPTION
**Pull request by nsingh-branch**
## Reference
SDK-2458 -- [iOS] Implement Consumer Protection Preferences

## Summary
Added a new method for setting the consumer protection preference, `setConsumerProtectionPreference`. This value can be set and changed at any time, but persists across sessions. This change also deprecated the `setTrackingDisabled()` method since  setting the preference to `BranchConsumerProtectionPreferenceTrackingDisabled` will perform the same logic.

## Motivation

To give clients more control over what end user information is tracked or used for attribution.

## Type Of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Testing Instructions

Call `[[Branch getInstance] setConsumerProtectionPreference:BranchConsumerProtectionPreferenceAnalyticsOnly];`, or any other preference, at any point in the app and then observe the respective field appearing in the following requests, like `"protection_preference" : 2,`

cc @BranchMetrics/saas-sdk-devs for visibility.

### CODE REVIEW

File: Sources/BranchSDK/Branch.m [hunk 1]
```diff
@@ -539,6 +539,42 @@ + (void) setDMAParamsForEEA:(BOOL)eeaRegion AdPersonalizationConsent:(BOOL)adPer
     [BNCPreferenceHelper sharedInstance].adUserDataUsageConsent = adUserDataUsageConsent;
 }
 
++ (void)setConsumerProtectionPreference:(BranchConsumerProtectionPreference)preference {
+    [BNCPreferenceHelper sharedInstance].consumerProtectionPreference = preference;
+    
+    [[BranchLogger shared] logVerbose:[NSString stringWithFormat:@"Setting Branch Consumer Protection Preference to %lu", (unsigned long)preference] error:nil];
```
Comment (gdeluna-branch) [inline Sources/BranchSDK/Branch.m:545]: Should probably remove company name in logging.

File: Sources/BranchSDK/Branch.m [hunk 2]
```diff
@@ -539,6 +539,42 @@ + (void) setDMAParamsForEEA:(BOOL)eeaRegion AdPersonalizationConsent:(BOOL)adPer
...
+
+    //Set tracking to disabled if consumer protection preference is changed to "No Attribution". Otherwise, keep tracking enabled.
+    if (preference == BranchConsumerProtectionPreferenceTrackingDisabled) {
+        if ([Branch trackingDisabled] == false) {
+            //Disable Tracking
+            [[BranchLogger shared] logVerbose:@"Disabling tracking due to No Attribution consumer protection preference." error:nil];
```
Comment (gdeluna-branch) [inline Sources/BranchSDK/Branch.m:551]: "Disabling attribution events..."

File: Sources/BranchSDK/Branch.m [hunk 3]
```diff
@@ -539,6 +539,42 @@ + (void) setDMAParamsForEEA:(BOOL)eeaRegion AdPersonalizationConsent:(BOOL)adPer
...
+            [branch clearNetworkQueue];
+            branch.initializationStatus = BNCInitStatusUninitialized;
+            [branch.linkCache clear];
+            // Release the lock in case it's locked:
+            [BranchOpenRequest releaseOpenResponseLock];
+        }
+    } else {
+        if ([Branch trackingDisabled]) {
+            //Enable Tracking
+            [[BranchLogger shared] logVerbose:@"Enabling tracking due to change in consumer protection preference." error:nil];
```
Comment (gdeluna-branch) [inline Sources/BranchSDK/Branch.m:568]: "Enabling attribution events..."

File: Sources/BranchSDK/BNCRequestFactory.m [hunk 4]
```diff
@@ -414,6 +414,10 @@ - (void)addAppClipDataToJSON:(NSMutableDictionary *)json {
 - (void)addDefaultRequestDataToJSON:(NSMutableDictionary *)json {
     json[@"branch_key"] = self.branchKey;
     
+    if (self.preferenceHelper.consumerProtectionPreference) {
```
Comment (gdeluna-branch) [inline Sources/BranchSDK/BNCRequestFactory.m:417]: Does this get retrieved in time for first open on install?

Comment (nsingh-branch) [reply Sources/BranchSDK/BNCRequestFactory.m]: I've confirmed that it does.

File: Sources/BranchSDK/Public/Branch.h [hunk 5]
```diff
@@ -812,6 +812,64 @@ Sets a custom base URL for all calls to the Branch API.
  */
 + (void) setDMAParamsForEEA:(BOOL) eeaRegion AdPersonalizationConsent:(BOOL) adPersonalizationConsent AdUserDataUsageConsent:(BOOL) adUserDataUsageConsent;
 
+
+/**
+ * Enumeration representing different levels of consumer protection attribution levels
+ */
+typedef NS_ENUM(NSUInteger, BranchAttributionLevel) {
```
Comment (gdeluna-branch) [inline Sources/BranchSDK/Public/Branch.h:819]: Same feedback, let's align on strings so there's no confusion in the future.

**Review by NidhiDixit09 — APPROVED** [review]
LGTM

**Pull request merged**